File: packages/android-kotlin/src/main/kotlin/com/onchainux/deeplink/DeepLinkHandler.kt
/**
 * DeepLinkHandler — handles deep linking for wallet connections on Android.
 *
 * Manages Intent-based deep links, universal links (App Links), and smart
 * redirect logic for Android wallet app connections.
 *
 * ## Usage
 * ```kotlin
 * val handler = DeepLinkHandler()
 * handler.configure(config)
 *
 * val intent = handler.generateIntent(walletId = "metamask", uri = wcUri)
 * context.startActivity(intent)
 * ```
 *
 * ## AndroidManifest.xml Setup
 * Add intent-filter for wallet deep links your app handles:
 * ```xml
 * <queries>
 *     <package android:name="io.metamask" />
 *     <package android:name="org.toshi" />
 *     <package android:name="me.rainbow" />
 *     <package android:name="com.wallet.crypto.trustapp" />
 * </queries>
 * ```
 */
package com.cinacoin.deeplink

import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.net.Uri
import com.cinacoin.core.CinacoinConfig

/** Platform for redirect decisions. */
enum class Platform { IOS, ANDROID, WEB }

/** Result of a redirect attempt. */
data class RedirectResult(
    val success: Boolean,
    val method: RedirectMethod,
    val url: String,
    val fallbackUsed: Boolean,
    val error: String? = null
)

/** Redirect method used. */
enum class RedirectMethod { DEEP_LINK, UNIVERSAL_LINK, APP_STORE, QR_CODE }

/** Parameters for deep link generation. */
data class DeepLinkParams(
    val walletId: String,
    val uri: String,
    val params: Map<String, String>? = null,
    val fallbackTimeoutMs: Int? = null
)

/** Configuration for a specific wallet's deep link scheme. */
data class WalletDeepLinkConfig(
    val scheme: String,
    val universalDomain: String? = null,
    val appStoreUrl: String? = null,
    val playStoreUrl: String? = null,
    val pathTemplate: String? = null,
    val packageName: String? = null
)

/**
 * Handles deep linking for wallet connections on Android.
 */
class DeepLinkHandler {

    companion object {
        /** Wallet deep link configurations. */
        val walletConfigs: MutableMap<String, WalletDeepLinkConfig> = mutableMapOf(
            "metamask" to WalletDeepLinkConfig(
                scheme = "metamask://",
                universalDomain = "metamask.app.link",
                playStoreUrl = "https://play.google.com/store/apps/details?id=io.metamask",
                pathTemplate = "/wc?uri={uri}",
                packageName = "io.metamask"
            ),
            "walletconnect" to WalletDeepLinkConfig(
                scheme = "wc://",
                universalDomain = "walletconnect.com",
                pathTemplate = "/wc?uri={uri}"
            ),
            "coinbase" to WalletDeepLinkConfig(
                scheme = "cbwallet://",
                universalDomain = "go.cb-w.com",
                playStoreUrl = "https://play.google.com/store/apps/details?id=org.toshi",
                packageName = "org.toshi"
            ),
            "rainbow" to WalletDeepLinkConfig(
                scheme = "rainbow://",
                universalDomain = "rnbwapp.com",
                playStoreUrl = "https://play.google.com/store/apps/details?id=me.rainbow",
                packageName = "me.rainbow"
            ),
            "trust" to WalletDeepLinkConfig(
                scheme = "trust://",
                universalDomain = "link.trustwallet.com",
                playStoreUrl = "https://play.google.com/store/apps/details?id=com.wallet.crypto.trustapp",
                packageName = "com.wallet.crypto.trustapp"
            ),
            "phantom" to WalletDeepLinkConfig(
                scheme = "phantom://",
                universalDomain = "phantom.app",
                playStoreUrl = "https://play.google.com/store/apps/details?id=com.phantom.app",
                packageName = "com.phantom.app"
            ),
            "rabby" to WalletDeepLinkConfig(
                scheme = "rabby://"
            )
        )
    }

    private var config: CinacoinConfig? = null

    /** Configure the handler. */
    fun configure(config: CinacoinConfig) {
        this.config = config
    }

    /**
     * Generate a deep link URL for a wallet.
     */
    fun generateDeepLink(walletId: String, uri: String, queryParams: Map<String, String>? = null): String {
        val walletConfig = walletConfigs[walletId] ?: return ""
        var url = walletConfig.scheme

        if (walletConfig.pathTemplate != null) {
            url += walletConfig.pathTemplate.replace("{uri}", uri)
        }

        if (!queryParams.isNullOrEmpty()) {
            val separator = if (url.contains("?")) "&" else "?"
            val query = queryParams.map { "${it.key}=${it.value}" }.joinToString("&")
            url += "$separator$query"
        }

        return url
    }

    /**
     * Generate an Intent for opening a wallet.
     */
    fun generateIntent(walletId: String, uri: String): Intent? {
        val url = generateDeepLink(walletId, uri)
        if (url.isEmpty()) return null

        val walletConfig = walletConfigs[walletId]
        val intent = Intent(Intent.ACTION_VIEW, Uri.parse(url))

        // Set package if known for direct app targeting
        walletConfig?.packageName?.let { pkg ->
            intent.setPackage(pkg)
        }

        intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
        return intent
    }

    /**
     * Open a wallet app using deep linking.
     */
    fun openWallet(context: Context, walletId: String, uri: String): RedirectResult {
        val intent = generateIntent(walletId, uri)
        if (intent == null) {
            return RedirectResult(false, RedirectMethod.QR_CODE, "", false, "Wallet not configured")
        }

        return try {
            context.startActivity(intent)
            RedirectResult(true, RedirectMethod.DEEP_LINK, intent.data.toString(), false)
        } catch (e: Exception) {
            // Try Play Store fallback
            val walletConfig = walletConfigs[walletId]
            if (walletConfig?.playStoreUrl != null) {
                val storeIntent = Intent(Intent.ACTION_VIEW, Uri.parse(walletConfig.playStoreUrl))
                storeIntent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
                try {
                    context.startActivity(storeIntent)
                    return RedirectResult(true, RedirectMethod.APP_STORE, walletConfig.playStoreUrl!!, true)
                } catch (_: Exception) {
                    // Fall through
                }
            }
            RedirectResult(false, RedirectMethod.DEEP_LINK, intent.data.toString(), false, e.message)
        }
    }

    /**
     * Generate a universal link URL (Android App Link).
     */
    fun generateUniversalLink(walletId: String, uri: String): String? {
        val walletConfig = walletConfigs[walletId]
        val domain = walletConfig?.universalDomain ?: return null
        val path = walletConfig.pathTemplate?.replace("{uri}", uri) ?: "/wc?uri=$uri"
        return "https://$domain$path"
    }

    /**
     * Check if a wallet app is installed.
     */
    fun isWalletInstalled(context: Context, walletId: String): Boolean {
        val walletConfig = walletConfigs[walletId]
        val packageName = walletConfig?.packageName ?: return false

        return try {
            context.packageManager.getPackageInfo(packageName, 0)
            true
        } catch (_: PackageManager.NameNotFoundException) {
            false
        }
    }

    /**
     * Register a custom wallet deep link configuration.
     */
    fun registerWallet(walletId: String, config: WalletDeepLinkConfig) {
        walletConfigs[walletId] = config
    }
}

File: packages/ios-swift/Sources/OnChainUX/DeepLinkHandler.swift
/**
 * DeepLinkHandler — handles deep linking and universal links for wallet connections.
 *
 * Manages URL scheme registration, deep link generation, and smart redirect
 * logic for iOS wallet app connections.
 *
 * ## Usage
 * ```swift
 * let handler = DeepLinkHandler()
 * handler.configure(with: config)
 *
 * // Generate a deep link
 * let url = handler.generateDeepLink(walletId: "metamask", uri: wcUri)
 *
 * // Open a wallet with fallback
 * let result = await handler.openWallet(walletId: "metamask", uri: wcUri)
 * ```
 *
 * ## Info.plist Setup
 * Add supported URL schemes for wallets you want to deep-link to:
 * ```xml
 * <key>LSApplicationQueriesSchemes</key>
 * <array>
 *     <string>metamask</string>
 *     <string>cbwallet</string>
 *     <string>rainbow</string>
 *     <string>trust</string>
 *     <string>phantom</string>
 * </array>
 * ```
 */

import Foundation

/// Platform for redirect decisions.
public enum DeepLinkPlatform: String, Sendable {
    case ios
    case android
    case web
}

/// Result of a deep link redirect attempt.
public struct RedirectResult: Sendable {
    /// Whether the redirect was initiated.
    public let success: Bool
    /// Method used: deep-link, universal-link, app-store.
    public let method: RedirectMethod
    /// The URL that was used.
    public let url: String
    /// Whether a fallback was used.
    public let fallbackUsed: Bool
    /// Error message if the redirect failed.
    public let error: String?
    
    public init(success: Bool, method: RedirectMethod, url: String, fallbackUsed: Bool, error: String? = nil) {
        self.success = success
        self.method = method
        self.url = url
        self.fallbackUsed = fallbackUsed
        self.error = error
    }
}

/// Redirect method used.
public enum RedirectMethod: String, Sendable {
    case deepLink = "deep-link"
    case universalLink = "universal-link"
    case appStore = "app-store"
    case qrCode = "qr-code"
}

/// Parameters for deep link generation.
public struct DeepLinkParams: Sendable {
    /// Wallet identifier.
    public let walletId: String
    /// URI to pass to the wallet.
    public let uri: String
    /// Additional query parameters.
    public let params: [String: String]?
    /// Timeout in ms before fallback.
    public let fallbackTimeoutMs: Int?
    
    public init(walletId: String, uri: String, params: [String: String]? = nil, fallbackTimeoutMs: Int? = nil) {
        self.walletId = walletId
        self.uri = uri
        self.params = params
        self.fallbackTimeoutMs = fallbackTimeoutMs
    }
}

/// Configuration for a specific wallet's deep link scheme.
public struct WalletDeepLinkConfig: Sendable {
    /// URL scheme prefix.
    public let scheme: String
    /// Universal link domain.
    public let universalDomain: String?
    /// App Store URL.
    public let appStoreUrl: String?
    /// Path template for the deep link.
    public let pathTemplate: String?
    
    public init(scheme: String, universalDomain: String? = nil, appStoreUrl: String? = nil, pathTemplate: String? = nil) {
        self.scheme = scheme
        self.universalDomain = universalDomain
        self.appStoreUrl = appStoreUrl
        self.pathTemplate = pathTemplate
    }
}

/// Handles deep linking for wallet connections.
public final class DeepLinkHandler {
    
    /// Wallet deep link configurations.
    public static var walletConfigs: [String: WalletDeepLinkConfig] = [
        "metamask": WalletDeepLinkConfig(
            scheme: "metamask://",
            universalDomain: "metamask.app.link",
            appStoreUrl: "https://apps.apple.com/app/metamask/id1438668043",
            pathTemplate: "/wc?uri={uri}"
        ),
        "walletconnect": WalletDeepLinkConfig(
            scheme: "wc://",
            universalDomain: "walletconnect.com",
            pathTemplate: "/wc?uri={uri}"
        ),
        "coinbase": WalletDeepLinkConfig(
            scheme: "cbwallet://",
            universalDomain: "go.cb-w.com",
            appStoreUrl: "https://apps.apple.com/app/coinbase-wallet/id1278383455"
        ),
        "rainbow": WalletDeepLinkConfig(
            scheme: "rainbow://",
            universalDomain: "rnbwapp.com",
            appStoreUrl: "https://apps.apple.com/app/rainbow-ethereum-wallet/id1457119021"
        ),
        "trust": WalletDeepLinkConfig(
            scheme: "trust://",
            universalDomain: "link.trustwallet.com",
            appStoreUrl: "https://apps.apple.com/app/trust-crypto-bitcoin-wallet/id1288339409"
        ),
        "phantom": WalletDeepLinkConfig(
            scheme: "phantom://",
            universalDomain: "phantom.app",
            appStoreUrl: "https://apps.apple.com/app/phantom-crypto-wallet/id1598432977"
        ),
        "rabby": WalletDeepLinkConfig(
            scheme: "rabby://"
        ),
    ]
    
    private var config: CinacoinConfig?
    
    /// Configure the handler.
    public func configure(with config: CinacoinConfig) {
        self.config = config
    }
    
    /// Generate a deep link URL for a wallet.
    /// - Parameters:
    ///   - walletId: Wallet identifier.
    ///   - uri: WalletConnect URI to pass.
    ///   - queryParams: Additional query parameters.
    /// - Returns: Deep link URL string.
    public func generateDeepLink(
        walletId: String,
        uri: String,
        queryParams: [String: String]? = nil
    ) -> String {
        guard let walletConfig = Self.walletConfigs[walletId] else {
            return ""
        }
        
        var url = walletConfig.scheme
        
        // Apply path template if available
        if let template = walletConfig.pathTemplate {
            url += template.replacingOccurrences(of: "{uri}", with: uri)
        }
        
        // Append additional query params
        if let queryParams = queryParams, !queryParams.isEmpty {
            let separator = url.contains("?") ? "&" : "?"
            let query = queryParams.map { "\($0)=\($1)" }.joined(separator: "&")
            url += "\(separator)\(query)"
        }
        
        return url
    }
    
    /// Generate a universal link URL for a wallet.
    /// - Parameters:
    ///   - walletId: Wallet identifier.
    ///   - uri: WalletConnect URI to pass.
    /// - Returns: Universal link URL string.
    public func generateUniversalLink(walletId: String, uri: String) -> String? {
        guard let walletConfig = Self.walletConfigs[walletId],
              let domain = walletConfig.universalDomain else {
            return nil
        }
        
        let path = walletConfig.pathTemplate?.replacingOccurrences(of: "{uri}", with: uri) ?? "/wc?uri=\(uri)"
        return "https://\(domain)\(path)"
    }
    
    /// Open a wallet using deep linking with fallback logic.
    /// Tries: deep link → timeout → universal link → App Store.
    /// - Parameters:
    ///   - walletId: Wallet identifier.
    ///   - uri: WalletConnect URI.
    ///   - fallbackTimeoutMs: Timeout before fallback in milliseconds.
    /// - Returns: Redirect result.
    public func openWallet(
        walletId: String,
        uri: String,
        fallbackTimeoutMs: Int = 1500
    ) async -> RedirectResult {
        guard let walletConfig = Self.walletConfigs[walletId] else {
            return RedirectResult(
                success: false,
                method: .qrCode,
                url: "",
                fallbackUsed: false,
                error: "Wallet '\(walletId)' not configured for deep linking"
            )
        }
        
        // Step 1: Try deep link
        let deepLinkUrl = generateDeepLink(walletId: walletId, uri: uri)
        if let url = URL(string: deepLinkUrl), UIApplication.shared.canOpenURL(url) {
            let opened = UIApplication.shared.open(url)
            if opened {
                return RedirectResult(
                    success: true,
                    method: .deepLink,
                    url: deepLinkUrl,
                    fallbackUsed: false
                )
            }
        }
        
        // Step 2: Attempt deep link anyway (iOS may handle it without canOpenURL)
        if let url = URL(string: deepLinkUrl) {
            let opened = UIApplication.shared.open(url)
            if opened {
                // Step 3: Fallback to universal link after timeout
                try? await Task.sleep(nanoseconds: UInt64(fallbackTimeoutMs) * 1_000_000)
                
                if let universalUrl = generateUniversalLink(walletId: walletId, uri: uri),
                   let url = URL(string: universalUrl) {
                    let fallbackOpened = UIApplication.shared.open(url)
                    return RedirectResult(
                        success: fallbackOpened,
                        method: .universalLink,
                        url: universalUrl,
                        fallbackUsed: true
                    )
                }
                
                // Step 4: App Store fallback
                if let appStoreUrl = walletConfig.appStoreUrl,
                   let url = URL(string: appStoreUrl) {
                    let storeOpened = UIApplication.shared.open(url)
                    return RedirectResult(
                        success: storeOpened,
                        method: .appStore,
                        url: appStoreUrl,
                        fallbackUsed: true
                    )
                }
                
                return RedirectResult(
                    success: false,
                    method: .deepLink,
                    url: deepLinkUrl,
                    fallbackUsed: true,
                    error: "All fallback methods failed"
                )
            }
        }
        
        return RedirectResult(
            success: false,
            method: .deepLink,
            url: deepLinkUrl,
            fallbackUsed: false,
            error: "Could not open deep link"
        )
    }
    
    /// Register a custom wallet deep link configuration.
    /// - Parameters:
    ///   - walletId: Wallet identifier.
    ///   - config: Deep link configuration.
    public static func registerWallet(walletId: String, config: WalletDeepLinkConfig) {
        walletConfigs[walletId] = config
    }
}
